pkg/apis/build/v1alpha1/buildstrategy.go

```diff
@@ -0,0 +1,57 @@
+// Copyright The Shipwright Contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	// NamespacedBuildStrategyKind indicates that the buildstrategy type has a namespaced scope.
+	NamespacedBuildStrategyKind BuildStrategyKind = "BuildStrategy"
+
+	// ClusterBuildStrategyKind indicates that buildstrategy type has a cluster scope.
+	ClusterBuildStrategyKind BuildStrategyKind = "ClusterBuildStrategy"
+)
+
+// BuildStrategySpec defines the desired state of BuildStrategy
+type BuildStrategySpec struct {
+	BuildSteps []BuildStep `json:"buildSteps,omitempty"`
+}
+
+// BuildStep defines a partial step that needs to run in container for
+// building the image.
+type BuildStep struct {
+	corev1.Container `json:",inline"`
+}
+
+// BuildStrategyStatus defines the observed state of BuildStrategy
+type BuildStrategyStatus struct {
+}
+
+// BuildStrategyKind defines the type of BuildStrategy used by the build.
+type BuildStrategyKind string
+
+// StrategyRef can be used to refer to a specific instance of a buildstrategy.
+// Copied from CrossVersionObjectReference: https://github.com/kubernetes/kubernetes/blob/169df7434155cbbc22f1532cba8e0a9588e29ad8/pkg/apis/autoscaling/types.go#L64
+type StrategyRef struct {
+	// Name of the referent; More info: http://kubernetes.io/docs/user-guide/identifiers#names
+	Name string `json:"name"`
+
+	// BuildStrategyKind indicates the kind of the buildstrategy, namespaced or cluster scoped.
+	Kind *BuildStrategyKind `json:"kind,omitempty"`
+
+	// API version of the referent
+	// +optional
+	APIVersion string `json:"apiVersion,omitempty"`
+}
+
+// BuildStrategyIfc defines the common elements of build strategies
+type BuildStrategyIfc interface {
```

Comment:
What `Ifc` mean? Interface?

Reply:
Yes, because we are kind of the ugly position that `BuildStrategy` is already is used. If this would be called `NamespacedBuildStrategy` the more simpler name would be free to use.

Reply:
I though about the same @HeavyWombat, maybe we should rename the Go type to `NamespacedBuildStrategy` while retaining the name of the custom resource definition. On the other hand, this could also be confusing if somebody comes in new.

I checked golang [Interface Naming](https://golang.org/doc/effective_go.html#interface-names) conventions. But, there is nothing for our use case.

Sometimes, like in C# if I remember correctly, a good practice for interface naming is `I` as prefix. Would mean `IBuildStrategy` in this case. My subjective opinion is that this is better, but that's just because I am used to seeing interfaces with such a prefix while I have never seen the `Ifc` suffix so far.

Reply:
From what I can gather from source in the internet, there is a tendency to either use the `- er` pattern, the `I` prefix or the `Interface` suffix combined with the Go idea of short is always better. The `Ifc` comes from a project I once worked with.

With all that being said, my preferences would be in descending order:
- `BuilderStrategy`
- `IBuildStrategy`
- `BuildStrategyInterface`

As much as I would like a rename of the other type names, it would create more confusion I guess.

Reply:
Maybe we can vote, but like, vote electors, count that for a couple of days and then decide on what has not won the popular vote. Something like that.

Reply:
The `er` recommendation I saw, my understanding is that it is for one-method interfaces. Anyway, the current choice as well as any of what you list above is fine with me. I am also okay with the rest of your code changes. Not approving yet to give others a chance to do a review.

Reply:
Yeah, the `-er` recommendation would be for interfaces with one function in mind. I just though this would be a nice workaround for the wording issue.